### PR TITLE
Enable cache flush after every -F iteration. Default : 0 (No cache flush)

### PR DESCRIPTION
**Pull request by saurabhAMD**
Enable instruction cache flush after every -F iteration. Default : 0 (No cache flush)

### CODE REVIEW

File: src/common.cu [hunk 1]
```diff
@@ -103,9 +103,46 @@ static int average = 1;
...
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t"
+                     "s_nop 0 \n\t" ::
+                         :);
```
Comment (nileshnegi) [inline src/common.cu:144]: are the no. of these NOPs specific to MI300x?
also, can these be replaced by something like `s_nop 0xf`?

Comment (saurabhAMD) [reply src/common.cu:144]: I have used this as a reference for this command: https://github.com/ROCm/hipBLASLt/blob/84e55ed4cdad47089d3eeac0d7def923f133368a/clients/benchmarks/client_sequence.cpp#L39

File: src/common.cu [hunk 2]
```diff
@@ -437,6 +474,13 @@ testResult_t startColl(struct threadArgs* args, ncclDataType_t type, ncclRedOp_t
     }
     #endif
 
+    if(enable_cache_flush > 0 && (enable_cache_flush==1 || ((iter % enable_cache_flush) == 0))) {
```
Comment (gilbertlee-amd) [inline src/common.cu]: if (enable_cache_flush > 0 && (iter % enable_cache_file == 0)) would suffice here.
1 divides everything.

File: src/common.cu [hunk 3]
```diff
@@ -1042,6 +1090,7 @@ int main(int argc, char* argv[]) {
             "[-G,--cudagraph <num graph launches>] \n\t"
             "[-C,--report_cputime <0/1>] \n\t"
 	    "[-O,--out_of_place <0/1>] \n\t"
+      "[-F,--cache_flush <0/1>] \n\t"
```
Comment (gilbertlee-amd) [inline src/common.cu]: This doesn't need to be just 0 or 1 though.

File: src/common.cu [hunk 4]
```diff
@@ -437,6 +474,13 @@ testResult_t startColl(struct threadArgs* args, ncclDataType_t type, ncclRedOp_t
     }
     #endif
 
+    if(enable_cache_flush > 0 && (enable_cache_flush==1 || ((iter % enable_cache_flush) == 0))) {
+      hipDeviceProp_t deviceProps;
```
Comment (gilbertlee-amd) [inline src/common.cu]: gpu_block3 shouldn't be queried / calculated per iteration to avoid unnecessary overhead

File: src/common.cu (hunk 1 above)
Comment (nileshnegi) [reply src/common.cu:144]: if you have a cache-flush verification mechanism, can you check if replacing these 16 `s_nop 0` with 1 `s_nop 0xf` is functionally similar?

**Review by gilbertlee-amd — CHANGES_REQUESTED** [review]

File: src/common.cu (hunk 2 above)
Comment (nileshnegi) [reply src/common.cu]: also, is the idea to silently not run cache flushing if iterations are not divisible by the number of cache flush iterations? should there be a warning message? or failure to run?

Comment (saurabhAMD) [reply src/common.cu]: The mod operation to enable cache flush after every -F iteration. Thus, I don't think any warning message is needed.

File: src/common.cu (hunk 4 above)
Comment (saurabhAMD) [reply src/common.cu]: Thank you for your comment. Updated your suggestion in the latest commit.

File: src/common.cu (hunk 3 above)
Comment (saurabhAMD) [reply src/common.cu]: Thank you for your comment. Updated your suggestion in the latest commit.

**Review by gilbertlee-amd — APPROVED** [review]

**Pull request merged**